### ISSUE TITLE
Support declarative format for`#up`

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#296]
## Summary

TLDR: the developer experience would be nicer if `#up` supported a declarative form in addition to an imperative form.

* Imperative form: array of low-level operations:`#up: [...op.#Op]`
* Declarative form: any object with its own nested array: `#up: { #up: [...op.#Op], _ }`

This is the same pattern as `op.#Load.from` for example.

## Context

Currently any value can be assigned a processing pipeline via the special definition `#up`. The value of this field must be an array of low-level pipeline operations. For example:

```
import (
    “dagger.io/dagger/op”
)

website: {
  // Array of low-level operations to deploy the website
  #up: [
    op.#FetchContainer & { ref: “alpine” },
    op.#Exec & { args: [“echo”, “hello world!”] },
  ]
}
```

This “array of operations” format is in the *imperative* style: it’s a list of operations to execute in certain order. The rest of the cue configuration is in the *declarative* style: it describes a state that we want.

When developing for Dagger, the imperative style is more advanced, and should be reserved for power users that need maximum control and power. The less it is necessary to drop to imperative mode, the better (in my experience).

One way to avoid dropping to imperative mode is using simple helper packages such as `dagger.io/docker`. For example the example above could be re-written as:

```
import (
  “dagger.io/alpine”
  “dagger.io/docker”

)

website: docker.#Container & {
    image: alpine.#Image
    command: “echo hello world”
}
```

Sometimes you want to embed a container in a higher-level object to avoid field collisions, for example:

```
import (
  “dagger.io/alpine”
  “dagger.io/docker”

)

website: {
    // website-specific field
    hostname: string

    ctr: docker.#Container & {
        image: alpine.#Image
        command: “echo hello \(hostname)”
    }
}
```

However, this pattern breaks if you want to export a value from the pipeline:

```
import (
  “dagger.io/alpine”
  “dagger.io/docker”

)

website: {
    // website-specific field
    hostname: string
    // url is computed

    // We can’t use docker.#Container here... because of export
    #up: [
        op.#FetchContainer & { ref: “alpine” },
        op.#Exec & { args: [“sh”, “-c”, “echo hello world > /url.txt”] },
        op.#Export & { source: “/url.txt” },
  ]
}
```

This problem could be fixed by supporting the following format:

```
import (
  “dagger.io/alpine”
  “dagger.io/docker”

)

website: {
    // website-specific field
    hostname: string

    // #up may be imperative or declarative. Here we use the declarative form.
    #up: docker.#Container & {
        image: alpine.#Image
        command: “echo hello \(hostname) > /url.txt”
        export: source: “/url.txt”
    }
}
```
